toolchain/check/type_completion.cpp

```diff
@@ -70,16 +72,143 @@ static auto NoteIncompleteNamedConstraint(
+  const auto& require_self_specific =
+      context.specifics().Get(require_generic.self_specific_id);
+  auto require_self_specific_args =
+      context.inst_blocks().Get(require_self_specific.args_id);
+  // The last argument of a `require` generic is always `Self`, as require can
+  // not have any parameters of its own, only enclosing parameters.
+  return require_self_specific_args.back();
+}
+
+static auto GetFacetTypeInSpecific(Context& context,
+                                   SemIR::SpecificId specific_id,
+                                   SemIR::InstId facet_type)
+    -> SemIR::FacetTypeId {
+  auto const_facet_type = SemIR::GetConstantValueInSpecific(
+      context.sem_ir(), specific_id, facet_type);
+  if (const_facet_type == SemIR::ErrorInst::ConstantId) {
+    return SemIR::FacetTypeId::None;
   }
+  auto facet_type_in_specific = context.insts().GetAs<SemIR::FacetType>(
+      context.constant_values().GetInstId(const_facet_type));
+  return facet_type_in_specific.facet_type_id;
+}
+
+static auto RequireCompleteFacetType(Context& context, SemIR::LocId loc_id,
+                                     const SemIR::FacetType& facet_type,
+                                     MakeDiagnosticBuilderFn diagnoser)
+    -> bool {
+  struct SpecificForRequires {
+    SemIR::SpecificId specific_id;
+    SemIR::RequireImplsBlockId requires_block_id;
+  };
+  llvm::SmallVector<SpecificForRequires> specifics;
+  int specifics_done = 0;
+
+  llvm::SmallVector<SemIR::FacetTypeId> work = {facet_type.facet_type_id};
+  while (!work.empty()) {
+    auto next_facet_type_id = work.pop_back_val();
+    const auto& facet_type_info = context.facet_types().Get(next_facet_type_id);
+
+    for (auto extends : facet_type_info.extend_constraints) {
+      auto interface_id = extends.interface_id;
+      const auto& interface = context.interfaces().Get(interface_id);
+      if (!interface.is_complete()) {
+        if (diagnoser) {
+          auto builder = diagnoser();
+          NoteIncompleteInterface(context, interface_id, builder);
+          builder.Emit();
+        }
+        return false;
+      }
+      if (interface.generic_id.has_value()) {
+        specifics.push_back(
+            {extends.specific_id, interface.require_impls_block_id});
+      }
+    }
+
+    for (auto extends : facet_type_info.extend_named_constraints) {
+      auto named_constraint_id = extends.named_constraint_id;
+      const auto& constraint =
+          context.named_constraints().Get(named_constraint_id);
+      if (!constraint.is_complete()) {
+        if (diagnoser) {
+          auto builder = diagnoser();
+          NoteIncompleteNamedConstraint(context, named_constraint_id, builder);
+          builder.Emit();
+        }
+        return false;
+      }
+      if (constraint.generic_id.has_value()) {
+        specifics.push_back(
+            {extends.specific_id, constraint.require_impls_block_id});
+      }
+    }
```

Comment:
Would be nice if we could find some way to avoid this duplication, maybe add a TODO?